### PR TITLE
Hook unread counter

### PR DESCRIPTION
**Pull request by gyunaev**
Implemented a hook unread counter, allowing to invoke a process eachtime the unread count changes. This is for post-1.8.1 release since it adds new strings.

### CODE REVIEW

**Comment by Abestanis** [comment]
Looks good at first glance, but I won't be able to test it until tomorrow.

> This is for post-1.8.1

What is the holdup for the release? Do you want to add something before releasing?

**Comment by gyunaev** [comment]
That´s fine, we are not in hurry. No, just waiting for your translations :)

**Comment by Abestanis** [comment]
> No, just waiting for your translations

Which translations? Do you mean the Spanish translations?

**Comment by gyunaev** [comment]
There are a bunch of pull reqs open with translations. Will we close them before 1.8.1?

**Comment by Abestanis** [comment]
No, I don't think it is worth waiting for them.
You can have a look at the update to the Russian translation if you want.

File: src/dialogsettings.ui [hunk 1]
```diff
@@ -373,6 +373,27 @@
...
+          <widget class="QLabel" name="label">
+           <property name="text">
+            <string>Unread change cmd:</string>
+           </property>
+          </widget>
+         </item>
+         <item>
+          <widget class="QLineEdit" name="leProcessRunOnCountChange">
+           <property name="toolTip">
+            <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;If not empty, this command will be invoked every time the unread counter changes (including when it becomes zero). It is invoked via shell, as-is, with %NEW% replaced by the new unread count value, and %OLD% replaced with the old unread count value (which may be the same as new).&lt;/p&gt;&lt;p&gt;Most users don't need this functionality and should leave it empty.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
```
Comment (Abestanis) [inline src/dialogsettings.ui:388]: I would remove the `Most users don't need this functionality and should leave it empty.`
Maybe it is a sign that this belongs in the `Advanced` tab? Or alternatively, leave it here but make it invisible unless enabled via a `Show advanced options` option in the `Advanced` tab?

File: src/dialogsettings.ui [hunk 2]
```diff
@@ -373,6 +373,27 @@
          </layout>
         </widget>
        </item>
+       <item>
+        <layout class="QHBoxLayout" name="horizontalLayout">
+         <item>
+          <widget class="QLabel" name="label">
+           <property name="text">
+            <string>Unread change cmd:</string>
```
Comment (Abestanis) [inline src/dialogsettings.ui:381]:
```suggestion
            <string>Unread change command:</string>
```
We always used command in user visible stings. I am obviously not a native speaker, but `command` sounds nicer than `cmd` to me.

File: src/trayicon.cpp [hunk 3]
```diff
@@ -129,6 +129,23 @@ void TrayIcon::unreadCounterUpdate( unsigned int total, QColor color )
...
+    // Execute the hook process
+    if ( !BirdtrayApp::get()->getSettings()->mProcessRunOnCountChange.isEmpty() )
+    {
+        QString cmdline = BirdtrayApp::get()->getSettings()->mProcessRunOnCountChange;
+
+        // Replace the %NEW% with the new unread count
+        cmdline.replace( "%NEW%", QString::number( total ) );
+
+        // Replace the %OLD% with the old unread count
+        cmdline.replace( "%OLD%", QString::number( mUnreadCounter ) );
```
Comment (Abestanis) [inline src/trayicon.cpp:141]: What do you think about adding %DIFF% as well? Should be really easy to add and might make it easier for some commands.

**Comment by Maboroshy** [comment]
> One observation though: When Birdtray is started and we detect unread emails, we also run the hook. This feels like it is undesired behavior, I guess many people want to use that hook to detect new mails and not the old ones that have been sitting in their Inbox for 5 months (definitely not talking about my Inbox here wink)

I want to use this feature to update the unread counter on the icon's badge at dock. So running the hook at start looks like a desired behavior for my use case. That way I can go without any counter storage between sessions. Maybe there can be an extra checkbox to make it optional.

**Comment by Maboroshy** [comment]
I've tested this branch on Manjaro Linux. All good. Managed to update Latte Dock badge counter with a tiny Python script I'm going to publish after this feature is released.

**Comment by rnc** [comment]
Is this for #304 ? 
Would it be possible to also pass e.g. account causing the count change ? 

Does the parser have knowledge of the folder (and even the sender/subject) as well which could be passed through?

**Comment by gyunaev** [comment]
Of folder, yes. But not of the account. Note that many files could have changed.

**Comment by rnc** [comment]
> Of folder, yes. But not of the account. Note that many files could have changed.

@gyunaev Thanks for the reply. Could I clarify that please - when I choose `Accounts to Monitor` it displays `Profile/Server(Account)/Folder` e.g. `Profile2020 / imap.gmail.com / Bugzilla `. Therefore surely if it knows about the Bugzilla folder, it should know about the imap.gmail.com account? 

Yes I understand its feasible given the timings that multiple emails may have arrived. What about something like sending to the external program 
```
<numericCountChange> [ <server>:<folder ].....n
```
 i.e. send a list of server , separator, folder as  a list of parameters ?

**Comment by gyunaev** [comment]
Too much work. You're welcome, of course, to implement it and submit a pull request.

**Comment by rnc** [comment]
@gyunaev Am I following it correctly with my supposition regarding server/folder knowledge?

Are you able to pass any further information in the interim? I haven't done C++ in decades so would have to see if I have any time to dig out my old books on it ;-)

**Comment by Abestanis** [comment]
> Am I following it correctly with my supposition regarding server/folder knowledge?

Yes, we do have that information. We will definitely be able to pass more information to the hook, but I agree with @gyunaev, let's first get the basics going. We will also need all of that information for #226 and #163. I did start working on that a few moths ago but wasn't able to finish it. Once I do we can add all of the information to the hook.

**Pull request merged**